### DISCUSSION TITLE
getting "Failed to initialize secure credential manager" while using profInfo.readProfilesFromDisk()

### DISCUSSION THREAD

**Discussion by kumnar02** [zowe/zowe-cli#2295] [Q&A] [answered]
using below code m i'm getting Failed to initialize secure credential manager , any idea whats wrong ?
```typescript

import { ProfileInfo } from "@zowe/imperative";
import { GetJobs } from "@zowe/zos-jobs-for-zowe-sdk";

(async () => {
    // Load connection info from default z/OSMF profile
    const profInfo = new ProfileInfo("zowe");
    await profInfo.readProfilesFromDisk();
    const zosmfProfAttrs = profInfo.getDefaultProfile("zosmf");
    const zosmfMergedArgs = profInfo.mergeArgsForProfile(zosmfProfAttrs, { getSecureVals: true });
    const session = ProfileInfo.createSession(zosmfMergedArgs.knownArgs);

    const owner: string = session.ISession.user;
    // This may take awhile...
    const response = await GetJobs.getJobsByOwner(session, owner);
    console.log(response);
})().catch((err) => {
    console.error(err);
    process.exit(1);
});

**Comment by kumnar02** [comment]
i have the .zowe/zowe.config.json file which is a team config

**Comment by t1m0thyj** [comment]
@kumnar02 Do you have the Zowe Secrets SDK ([`@zowe/secrets-for-zowe-sdk`](https://www.npmjs.com/package/@zowe/secrets-for-zowe-sdk)) installed in your project? You need this package in order to load secure credentials from the OS vault.

If you still have issues after installing the Secrets SDK package, please provide your OS, Node version, and npm version.

**Reply by kumnar02** [reply]
yes i have that installed and i still see issues but when i comment the below code it works exactly how i need.  its failing after this and going to catch block.

```
  const requireOpts = {};
                if (require.main.filename != null) {
                    requireOpts.paths = [require.main.filename, ...require.resolve.paths("@zowe/secrets-for-zowe-sdk")];

```

```typescript

initialize() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                console.log(require.main);
 
                // Imperative overrides the value of require.main.filename to point to
                // our calling CLI. Since our caller must supply keytar, we search for keytar
                // within our caller's path.
                /* 
                const requireOpts = {};
                if (require.main.filename != null) {
                    requireOpts.paths = [require.main.filename, ...require.resolve.paths("@zowe/secrets-for-zowe-sdk")];
                } */
                // use helper function for require.resolve so it can be mocked in jest tests
                const keytarPath = require.resolve("@zowe/secrets-for-zowe-sdk");
                logger_1.Logger.getImperativeLogger().debug("Loading Keytar module from", keytarPath);
                this.keytar = (yield Promise.resolve(`${keytarPath}`).then(s => require(s))).keyring;
            }
            catch (error) {
                this.loadError = new error_1.ImperativeError({
                    msg: `Failed to load Keytar module: ${error.message}`,
                    causeErrors: error
                });
                logger_1.Logger.getImperativeLogger().debug("Failed to load Keytar module:\n", error.stack);
            }
        });
    }

**Reply by kumnar02** [reply]
@t1m0thyj , can you check on this and let me know what's wrong here ?

**Reply by traeok** [reply]
Hi @kumnar02, I have a couple follow-up questions:

- Are you using Webpack or a bundler for your project?
- Can you provide the full error that you encountered? The error message might uncover additional information needed to resolve your issue.

---

To verify whether the Secrets SDK can be successfully resolved in your project, you can spawn an interactive shell within the root of your project folder and try to import it using `require`. If it imports without error and you can successfully call one of its functions (`findCredentials` used in the example below), then the Secrets SDK is properly configured for your project.
```
> node -i

Welcome to Node.js v20.13.1.
Type ".help" for more information.
> const { keyring } = require("@zowe/secrets-for-zowe-sdk")
undefined
> await keyring.findCredentials("Zowe", "*")
[
  {
    account: 'secure_config_props',
    password: 'eyJDOlxcVXNlcn... (rest of password)'
  }
]
```

If you run into any issues when following the above example, ensure that the Secrets SDK is specified as a normal dependency of your project and not a dev or peer dependency.

Hope this helps!

**Reply by kumnar02** [reply]
@traeok  Please find my answers below 
Are you using Webpack or a bundler for your project? - unbundled.

tried like above and it worked.
![success](https://github.com/user-attachments/assets/63b968d0-e937-43a0-b41b-473fd754118c)

my strong guess is it is failing and the error is from(DefaultCredentialManager.ts) while retrieving this require.main.filename and it goes to catch block

const requireOpts = {};
                if (require.main.filename != null) {
                    requireOpts.paths = [require.main.filename, ...require.resolve.paths("@zowe/secrets-for-zowe-sdk")];
                }

**Reply by traeok** [reply, marked as answer]
Thanks for the clarification @kumnar02 - this seems to be an oversight with the logic in `DefaultCredentialManager`. In some environments, the `main` property is not always available in the `require` object, so the `if` check that you posted results in a TypeError.

I've pushed up a fix for this and hoping to get it merged in as soon as possible. Thanks for bringing this issue to our attention.

**Reply by traeok** [reply]
Hi @kumnar02, we've released a fix for this bug and it will be available in `@zowe/imperative@8.1.2`. Feel free to follow up here if you have any questions or concerns when using the new version.

_Correction:_ The next version will be 8.1.2, not 8.1.1 - sorry for the confusion.

**Reply by kumnar02** [reply]
Perfect thank you.. that was a quick solution.. I will give it a try...
